File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.py
def solution(id_list, report, k):
    report_map = {id: set() for id in id_list}
    reported_map = {id: set() for id in id_list}
    
    for report_info in report:
        report_user, reported_user = report_info.split()
        report_map[report_user].add(reported_user)
        reported_map[reported_user].add(report_user)
    
    answer = []
    for id in id_list:
        count = 0
        for reported_user in report_map[id]:
            if len(reported_map[reported_user]) >= k:
                count += 1
        answer.append(count)
    
    return answer

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.*;

class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        Map<String, Set<String>> userReportMap = new HashMap<>();
        Map<String, Set<String>> reportedMap = new HashMap<>();
        
        for (String id : id_list) {
            userReportMap.put(id, new HashSet<>());
            reportedMap.put(id, new HashSet<>());
        }
        
        
        for (String reportInfo : report) {
            String[] reportArr = reportInfo.split(" ");
            String reportUser = reportArr[0];
            String reportedUser = reportArr[1];
            
            userReportMap.get(reportUser).add(reportedUser);
            reportedMap.get(reportedUser).add(reportUser);
        }
        
        int[] answer = new int[id_list.length];
        
        for (int i = 0; i < id_list.length; i++) {
            // report user의 key값으로 set을 만듬
            // set에는 신고당한 리스트가 들어있음
            Set<String> set = userReportMap.get(id_list[i]);
            int count = 0;
            for (String reportedUser : set) {
                if (reportedMap.get(reportedUser).size() >= k) {
                    count++;
                }
            }
            answer[i] = count;
        }
        
        return answer;
    }
}